src/poly/evaluator.rs

```diff
@@ -0,0 +1,542 @@
+pub(crate) fn new_evaluator<E: Fn() + Clone, F: Field, B: Basis>(context: E) -> Evaluator<E, F, B> {
+    Evaluator {
+        polys: vec![],
+        _context: context,
+    }
+}
+
+impl<E, F: Field, B: Basis> Evaluator<E, F, B> {
+    /// Registers the given polynomial for use in this evaluation context.
+    ///
+    /// This API treats each registered polynomial as unique, even if the same polynomial
+    /// is added multiple times.
+    pub(crate) fn register_poly(&mut self, poly: Polynomial<F, B>) -> AstLeaf<E, B> {
+        let index = self.polys.len();
+        self.polys.push(poly);
+
+        AstLeaf {
+            index,
+            rotation: Rotation::cur(),
+            _evaluator: PhantomData::default(),
+        }
+    }
+
+    /// Evaluates the given polynomial operation against this context.
+    pub(crate) fn evaluate(
+        &self,
+        ast: &Ast<E, F, B>,
+        domain: &EvaluationDomain<F>,
+    ) -> Polynomial<F, B>
+    where
+        E: Copy + Send + Sync,
+        F: FieldExt,
+        B: BasisOps,
+    {
+        // Traverse `ast` to collect the used leaves.
+        fn collect_rotations<E: Copy, F: Field, B: Basis>(
+            ast: &Ast<E, F, B>,
+        ) -> HashSet<AstLeaf<E, B>> {
+            match ast {
+                Ast::Poly(leaf) => vec![*leaf].into_iter().collect(),
+                Ast::Add(a, b) | Ast::Mul(AstMul(a, b)) => {
+                    let lhs = collect_rotations(a);
+                    let rhs = collect_rotations(b);
+                    lhs.union(&rhs).cloned().collect()
+                }
+                Ast::Scale(a, _) => collect_rotations(a),
+                Ast::LinearTerm(_) | Ast::ConstantTerm(_) => HashSet::default(),
+            }
+        }
+        let leaves = collect_rotations(ast);
+
+        // Produce the rotated polynomials.
+        let rotated: HashMap<_, _> = leaves
+            .iter()
+            .cloned()
+            .map(|leaf| {
+                (
+                    leaf,
+                    B::rotate(domain, &self.polys[leaf.index], leaf.rotation),
+                )
+            })
+            .collect();
+
+        // We're working in a single basis, so all polynomials are the same length.
+        let poly_len = self.polys.first().unwrap().len();
+        // Check the level of parallelization we have available.
+        let num_threads = multicore::current_num_threads();
+        // We scale the number of chunks by a constant factor, to ensure that if not all
+        // threads are available, we can achieve more uniform throughput and don't end up
+        // waiting on a couple of threads to process the last chunks.
+        let num_chunks = num_threads * 4;
+        // Calculate the ideal chunk size for the desired throughput. We use ceiling
+        // division to ensure the minimum chunk size is 1.
+        //     chunk_size = ceil(poly_len / num_chunks)
+        let chunk_size = (poly_len + num_chunks - 1) / num_chunks;
```

Comment:
We probably want to impose a minimum greater than 1; otherwise the cost of traversing the AST will be significant compared to doing real work on the polynomial chunks. (This mitigates the possibility of having selected too many chunks above on machines that have many threads, with a smallish circuit.)

Reply:
Yep; again, let's do that in tuning.